### PR TITLE
ranger: add a more common method to calc range.

### PR DESCRIPTION
**Pull request by winoros**
This method use []*expression.Column as parameter and both IntRange, ColumnRange, IndexRange can use this method.
@hanfei1991 PTAL

### CODE REVIEW

File: util/types/range.go [hunk 1]
```diff
@@ -22,6 +22,11 @@ import (
 	"github.com/pingcap/tidb/sessionctx/variable"
 )
 
+// Range is the interface of the three type of range.
```
Comment (hanfei1991) [inline util/types/range.go]: 1. I think this file can be moved to package util/ranger
2. I think this interface can implement convert2***Range

File: expression/column.go [hunk 2]
```diff
@@ -245,3 +245,43 @@ func Column2Exprs(cols []*Column) []Expression {
...
+	}
+	for _, c := range cols {
+		if c.ColName.L == col.Name.L {
+			return c
+		}
+	}
+	return nil
+}
+
+// indexCol2Col finds the correspond column of the IndexColumn in a column slice.
```
Comment (hanfei1991) [inline expression/column.go]: corresponding

File: expression/column.go [hunk 3]
```diff
@@ -245,3 +245,43 @@ func Column2Exprs(cols []*Column) []Expression {
 	}
 	return result
 }
+
+// ColInfo2Col finds the correspond column of the ColumnInfo in a column slice.
```
Comment (hanfei1991) [inline expression/column.go]: corresponding

File: util/types/range.go (hunk 1 above)
Comment (winoros) [inline util/types/range.go]: 1st comment make the pr a little large, move this to a new pr.
2nd done.

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -820,18 +826,23 @@ func (p *DataSource) convertToTableScan(prop *requiredProp) (task task, err erro
...
 	if len(p.pushedDownConds) > 0 {
 		conds := make([]expression.Expression, 0, len(p.pushedDownConds))
 		for _, cond := range p.pushedDownConds {
 			conds = append(conds, cond.Clone())
 		}
-		ts.AccessCondition, ts.filterCondition = ranger.DetachColumnConditions(conds, p.tableInfo.GetPkName())
-		ts.Ranges, err = ranger.BuildTableRange(ts.AccessCondition, sc)
-		if err != nil {
-			return nil, errors.Trace(err)
+		if pkColumn != nil {
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: the `GetPkColInfo` may return nil, it's not a good code style. Here we can check pk is handle and make sure pk column is not nil.

File: util/ranger/new_refiner.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	ranges := rb.buildColumnRanges(rangePoints, tp)
+	if rb.err != nil {
+		return nil, errors.Trace(rb.err)
+	}
+	return ranges, nil
+}
+
+// BuildRange is a method which can calculate IntColumnRange, ColumnRange, IndexRange.
+func BuildRange(sc *variable.StatementContext, conds []expression.Expression, rangeType int, cols []*expression.Column, lengths []int) (retRanges []types.Range,
```
Comment (hanfei1991) [inline util/ranger/new_refiner.go]: I think it's no need to pass every column's length. It's enough to know whether the last column has specified length. e.g. if the index is (a,b,c) but b has a length, we can only pass the columns (a,b)

Comment (winoros) [inline util/ranger/new_refiner.go]: Every specified length is used. So we should pass them.

File: model/model.go [hunk 6]
```diff
@@ -135,6 +135,18 @@ func (t *TableInfo) GetPkName() CIStr {
 	return CIStr{}
 }
 
+// GetPkColInfo gets the ColumnInfo of pk if exists.
+func (t *TableInfo) GetPkColInfo() *ColumnInfo {
+	if t.PKIsHandle {
```
Comment (hanfei1991) [inline model/model.go]: This check can be removed

File: util/ranger/new_refiner.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	}
+	return
+}
+
+// RangeSlice2IntRangeSlice changes []types.Range to []types.IntColumnRange
+func RangeSlice2IntRangeSlice(ranges []types.Range) []types.IntColumnRange {
```
Comment (hanfei1991) [inline util/ranger/new_refiner.go]: This function's name is too long. I think Ranges2IntRanges is ok. The following are same.

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by winoros** [comment]
@shenli PTAL

File: model/model.go [hunk 8]
```diff
@@ -135,6 +135,16 @@ func (t *TableInfo) GetPkName() CIStr {
 	return CIStr{}
 }
 
+// GetPkColInfo gets the ColumnInfo of pk if exists.
+func (t *TableInfo) GetPkColInfo() *ColumnInfo {
+	for _, colInfo := range t.Columns {
+		if mysql.HasPriKeyFlag(colInfo.Flag) {
```
Comment (shenli) [inline model/model.go]: How about composed pk？

Comment (winoros) [inline model/model.go]: composed pk currently is treated as a index.

Comment (winoros) [inline model/model.go]: TableInfo.PkIsHandle checked outside when call this method. Move it in here?

Comment (shenli) [inline model/model.go]: If the logic is right, you can keep it untouched. But please add a comment here.

File: expression/column.go [hunk 9]
```diff
@@ -245,3 +245,40 @@ func Column2Exprs(cols []*Column) []Expression {
...
+func indexCol2Col(cols []*Column, col *model.IndexColumn) *Column {
+	for _, c := range cols {
+		if c.ColName.L == col.Name.L {
+			return c
+		}
+	}
+	return nil
+}
+
+// IndexInfo2Cols changes a indexInfo's []*IndexColumn to corresponding []*Column and the length of each *IndexColumn.
```
Comment (shenli) [inline expression/column.go]: The first sentence is not clear.

**Comment by shenli** [comment]
LGTM

**Pull request merged**